dev/com.ibm.ws.cdi.1.2.weld/src/com/ibm/ws/cdi/proxy/ProxyServicesImpl.java

```diff
@@ -132,5 +139,32 @@ protected void addWeldDynamicImports(Bundle b, ManifestElement[] dynamicImports)
             throw new RuntimeException(e.getTargetException());
         }
     }
+
+    private void stackCheck() {
+        if (System.getSecurityManager() != null) {
+            Class<?>[] stackClasses = AccessController.doPrivileged(new PrivilegedAction<Class<?>[]>() {
+                @Override
+                public Class<?>[] run() {
+                    return new SecurityManager() {
+                        @Override
+                        protected Class<?>[] getClassContext() {
+                            return super.getClassContext();
+                        }
+                    }.getClassContext();
+                }
+            });
+
+            for (int i = 0; i < stackClasses.length; i++) {
+                if (! stackClasses[i].equals(this.getClass())) {
+                    continue;
+                }
+                if (stackClasses[i].equals(org.jboss.weld.bean.proxy.ProxyFactory.class)) {
+                    break;
+                } else {
+                    throw new AccessControlException("Someone other than weld attempted to access proxy services");
+                }
+            }
```

Comment:
This logic doesn't look right.

```java
                if (! stackClasses[i].equals(this.getClass())) {
                    continue;
                }
```
Doesn't this skip examining any class that isn't the current class? I think the `!` should be removed?

Do we have tests for this? This code looks like it should fail in all cases - the first thing on the stack would be this class so the check on 158 would fail, then the check on 161 would fail because this class is not `ProxyFactory` and we'd throw an `AccessControlException`?

Reply:
I've been using the passivation tests to test for this since the weld docs suggested they'd trigger this method. Clearly they don't. Now I wonder if anything we're doing goes through here.

Reply:
We should be able to tell from trace whether it does or not.